src/ingredient.rs

```diff
@@ -23,10 +23,19 @@ pub trait Jar: Any {
         aux: &dyn JarAux,
         first_index: IngredientIndex,
     ) -> Vec<Box<dyn Ingredient>>;
+
+    /// If this jar's first ingredient is a salsa struct, return its `TypeId`
+    fn salsa_struct_type_id(&self) -> Option<TypeId>;
 }
 
 pub trait JarAux {
-    fn next_memo_ingredient_index(&self, ingredient_index: IngredientIndex) -> MemoIngredientIndex;
+    fn lookup_jar_by_type(&self, jar: &dyn Jar) -> Option<IngredientIndex>;
```

Comment:
Pre-existing, but we should add comments to these methods. What is the reason people would invoke `lookup_jar_by_type` and `next_memo_ingredient_index`?

Reply:
```suggestion
    /// Return index of first ingredient from `jar` (based on the dynamic type of `jar`).
    /// Returns `None` if the jar has not yet been added. 
    /// Used by tracked functions to lookup the ingredient index for the salsa struct they take as argument.
    fn lookup_jar_by_type(&self, jar: &dyn Jar) -> Option<IngredientIndex>;
```